### PR TITLE
[KOGITO-7253] Workflow error handling guide

### PR DESCRIPTION
**Pull request by fjtirado**
**JIRA:**
https://issues.redhat.com/browse/KOGITO-7253

### CODE REVIEW

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 1]
```diff
@@ -1 +1,77 @@
-//= Understanding workflow error handling
\ No newline at end of file
+= Understanding workflow error handling
+
+This guide explains how to handle errors that might occur while executing workflow states. Although a non desirable situation, Workflow states, as part of their interaction with external systems, might fail. For these situations, Serverless Workflow specification provides an link:{spec_doc_url}#workflow-error-handling[error handling] mechanism which allows users to deal with these rainy day circumstances.
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:3]: I didn't understand the first part correctly.
Handle errors is how to handle errors or handle errors is a method from serverless workflow? (a question that could be asked)

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 2]
```diff
@@ -1 +1,77 @@
...
+
+An error is an exceptional occurrence that alters the regular Workflow execution sequence. When an error occurs, rather than transitioning to the state that would have been executed in case of a successful execution, the flow, if configured to do so, jumps instead to an alternative state, potentially capable of dealing with it. Note that error definition is always optional, if no error handling is defined, the Workflow execution will be aborted when something goes wrong. If you are a software developer, you can can think of Workflow error handling  as a `try-catch` or a `goto` construct.
+
+== Error definition
+
+An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: “wrong parameter”, while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of an link:https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html[Exception type] that might be thrown by the state.
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:9]:
```suggestion
An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: 'wrong parameter', while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of an link:https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html[Exception type] that might be thrown by the state.
```

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 3]
```diff
@@ -1 +1,77 @@
...
+
+The algorithm used by Kogito to determine if an exception should be handled by an error definition is:
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:11]: shouldn't be steps instead algorithm?

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 4]
```diff
@@ -1 +1,77 @@
...
+
+* it checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition.
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:13]:
```suggestion
* It checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 
```

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 5]
```diff
@@ -1 +1,77 @@
...
+
+== Error definition
+
+An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: “wrong parameter”, while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of an link:https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html[Exception type] that might be thrown by the state.
+
+The algorithm used by Kogito to determine if an exception should be handled by an error definition is:
+
+* it checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 
+* if this not the case, it recursively checks if the root cause of the Java Exception is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition.
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:14]: checks if the root cause.... is duplicated from the first bullet, maybe rephrase it?

**Review by spolti — CHANGES_REQUESTED** [review]
Nice work, just a few comments.

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc (hunk 3 above)
Comment (fjtirado) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:11]: Well, al algorithm is another name for a set of rules, isnt it? ;)

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc (hunk 5 above)
Comment (fjtirado) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc]: yes, I normally dislike duplication, but in this case,  rephrasing the second bullet might indicate the procedure is different for the root cause, which is not. 
What about "if this not the case, it recursively performs the previous checks over the the chains of root causes of the Java Exception"?

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 6]
```diff
@@ -1 +1,77 @@
...
+
+== Error handling example 
+
+Lets illustrate error handling with this link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus[Serverless Workflow Application].
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:18]: Too many lets, isnt it? I think the fact the diagram will appear below this paragraph make this lets take a look kind of implicit, wdyt?

[the remark this answers is not in the document]
Comment (spolti) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:18]: maybe, yes.

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 7]
```diff
@@ -1 +1,77 @@
...
+[source, json]
+----
+"errors": [
+    {
+      "name": "odd number",
+      "code": "java.lang.RuntimeException"
+     }
+  ]
+----
+* Referring this error definition in the 'checkEven' state to tell the workflow that if this error occurs it should transition to `odd` state rather than the regular transition to `even` state. Since `IllegalArgumentException` is a child of `RuntimeException`, it will be handled by the `errorRef` if thrown by `isEven` function.
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:47]:
```suggestion
* Referring to this error definition in the 'checkEven' state to tell the workflow that if this error occurs, it should transition to an `odd` state rather than the regular transition to the `even` state. Since `IllegalArgumentException` is a child of `RuntimeException`, it will be handled by the `errorRef` if thrown by the `isEven` function.
```

**Review by ricardozanini — COMMENTED** [review]
Great work! I've added a few minor grammar fixes, but I think Heena will handle this in the future.

I'm not concerned about this guide, but users might find it difficult to map an error within the workflow to a Java Exception without knowing the internals or having Java knowledge. I've opened https://issues.redhat.com/browse/KOGITO-7453 to have a more contextual approach. This is just the beginning, and we can iterate on this later.

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc (hunk 3 above)
Comment (fjtirado) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:11]: ok, changed to steps

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc (hunk 5 above)
Comment (fjtirado) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc]: changed, please take a look @spolti

**Review by ricardozanini — APPROVED** [review]

**Review by spolti — APPROVED** [review]

**Review by hmanwani-rh — APPROVED** [review]
@fjtirado I've added tweaks to the content and also fixed the image. Please take a look at the applied changes and let me know if we're good to merge this PR. 

Doc preview: https://deploy-preview-31--tangerine-crostata-04a6bf.netlify.app/serverlessworkflow/main/core/understanding-workflow-error-handling.html

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 8]
```diff
@@ -7,31 +7,43 @@ Understanding workflow error handling
 // links
 :java_exception_url: https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html
 
-This guide explains how to handle errors that might occur while executing workflow states. Although a non-desirable situation, Workflow states might fail as part of their interaction with external systems. For these situations, the Serverless Workflow specification provides an link:{spec_doc_url}#workflow-error-handling[error handling] mechanism, allowing users to deal with these rainy day circumstances.
+This document describes how you can handle the errors that might occur when you execute your Serverless Workflow states.
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:10]: Not sure if "when you execute your Serverless Workflow states" passes the right message. It's not the user who executes the states; it's the engine. Too bad if left as "when the Serverless Workflow states are executed"? I know that's passive voice, but it makes more sense to me.

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc [hunk 9]
```diff
@@ -7,31 +7,43 @@ Understanding workflow error handling
...
 
-The workflow in the link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus[Serverless Workflow Error Quarkus] example illustrates the error handling feature, as shown in the image below. 
+.Example of error handling
+image::core/error_handling.png[]
 
-image::core/error_handling.png []
+In the `serverless-workflow-error-quarkus` example application, the workflow accepts an integer number as an input model. When the workflow starts, it invokes the `isEven` Java service, which passes the the defined integer number as a parameter. The link:{kogito_sw_examples_url}serverless-workflow-error-quarkus/src/main/java/org/kie/kogito/examples/EvenService.java[`isEven` Java service] throws an `IllegalArgumentException` exception if the passed integer number is odd. This means that if the integer number is even, the Serverless Workflow transitions to the even state, however, if the integer number is odd, the Serverless Workflow transitions to the odd state, resulting in a different model output.
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:38]:
```suggestion
In the `serverless-workflow-error-quarkus` example application, the workflow accepts an integer number as an input model. When the workflow starts, it invokes the `isEven` Java service, which passes the defined integer number as a parameter. The link:{kogito_sw_examples_url}serverless-workflow-error-quarkus/src/main/java/org/kie/kogito/examples/EvenService.java[`isEven` Java service] throws an `IllegalArgumentException` exception if the passed integer number is odd. This means that if the integer number is even, the Serverless Workflow transitions to the `even` state, however, if the integer number is odd, the Serverless Workflow transitions to the `odd` state, resulting in a different model output.
```

Small change just to highlight that even and odd, in this context, are state names.

**Review by ricardozanini — APPROVED** [review]
@hmanwani-rh @fjtirado I've left a few minor suggestions to the last review.

File: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc (hunk 9 above)
Comment (fjtirado) [reply serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc]: I already changed

**Comment by fjtirado** [comment]
> @fjtirado I've added tweaks to the content and also fixed the image. Please take a look at the applied changes and let me know if we're good to merge this PR.
> 
@hmanwani-rh There were some things you changed that alter the original meaning, so I have to rewrite again. Next time, please, rather than directly comitting your changes, open a PR over the PR and lets discuss your changes over there.

**Comment by ricardozanini** [comment]
I'm merging this PR. Please if we have more divergences, open follow-up patch PR to discuss there.
